contracts/PrepaidCardMarketV2.sol

```diff
@@ -0,0 +1,361 @@
+pragma solidity ^0.8.9;
+pragma abicoder v1;
+
+import "./core/Ownable.sol";
+
+import "@openzeppelin/contracts-upgradeable/utils/structs/EnumerableSetUpgradeable.sol";
+import "@openzeppelin/contracts-upgradeable/token/ERC20/utils/SafeERC20Upgradeable.sol";
+import "./core/Versionable.sol";
+import "./IPrepaidCardMarket.sol";
+import "./PrepaidCardManager.sol";
+import "./ActionDispatcher.sol";
+import "./VersionManager.sol";
+import "./TokenManager.sol";
+import "./libraries/SafeERC677.sol";
+import "@openzeppelin/contracts-upgradeable/security/ReentrancyGuardUpgradeable.sol";
+
+contract PrepaidCardMarketV2 is
+  Ownable,
+  Versionable,
+  ReentrancyGuardUpgradeable
+{
+  using EnumerableSetUpgradeable for EnumerableSetUpgradeable.AddressSet;
+  using SafeERC20Upgradeable for IERC677;
+  using SafeERC677 for IERC677;
+
+  struct SKU {
+    address issuerSafe;
+    address issuer;
+    address issuingToken;
+    uint256 faceValue;
+    string customizationDID;
+  }
+
+  address public prepaidCardManagerAddress;
+  address public tokenManager;
+  address public provisioner;
+  address public actionDispatcher;
+  address public exchangeAddress;
+  address public versionManager;
+  mapping(address => mapping(address => uint256)) public balance; // issuer safe address -> token -> balance
+  mapping(address => address) public issuers; // issuer safe address -> issuer EOA
+  mapping(bytes32 => SKU) public skus; // sku => sku data
+  mapping(bytes32 => uint256) public asks; // sku => ask price (in issuing token)
+
+  bool public paused;
+
+  EnumerableSetUpgradeable.AddressSet internal trustedProvisioners;
+
+  event Setup();
+
+  event TokensDeposited(
+    address issuer,
+    uint256 amount,
+    address token,
+    address safe
+  );
+  event TokensWithdrawn(
+    address safe,
+    address issuer,
+    address token,
+    uint256 amount
+  );
+  event SkuAdded(
+    address issuer,
+    address issuingToken,
+    uint256 faceValue,
+    string customizationDID,
+    bytes32 sku
+  );
+  event AskSet(
+    address issuer,
+    address issuingToken,
+    bytes32 sku,
+    uint256 askPrice
+  );
+  event PrepaidCardProvisioned(address owner, bytes32 sku);
```

Comment:
Will `sku` be something we can use to get the address of the newly-provisioned prepaid card? I’m looking at reacting to the `PrepaidCardProvisioned` event, we’d like to be able to have the card’s address.

Reply:
no it's an entirely different thing. the address of the prepaid card is part of the event that the PrepaidCardManager emits in this same txn. This will be queryable from the subgraph based on the reservation table's provisioning txn hash. However we still need to do the subgraph work to capture the new events from the V2 market contract before we can tell you what the query is.